task/mgmt-gateway/src/update_buffer.rs

```diff
@@ -0,0 +1,148 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::{Log, __RINGBUF};
+use core::marker::PhantomData;
+use gateway_messages::ResponseError;
+use ringbuf::ringbuf_entry;
+use userlib::UnwrapLite;
+
+/// `UpdateBuffer` provides common logic for apply updates over the management
+/// network, assuming a common pattern of:
+///
+/// 1. A fixed block size.
+/// 2. Update size is known when the update begins, but data arives in
+///    arbitrarily-sized chunks.
+/// 3. A function to call for each complete block (and the final, potentially
+///    short block).
+/// 4. A function to call once all blocks have been written.
+pub struct UpdateBuffer<T, const BLOCK_SIZE: usize> {
+    total_length: usize,
+    bytes_written: usize,
+    current_block: &'static mut heapless::Vec<u8, BLOCK_SIZE>,
+    update_in_progress: bool,
+    write_block_fn: fn(&T, usize, &[u8]) -> Result<(), ResponseError>,
+    finalize_fn: fn(&T) -> Result<(), ResponseError>,
+    marker: PhantomData<fn() -> T>,
+}
+
+impl<T, const BLOCK_SIZE: usize> UpdateBuffer<T, BLOCK_SIZE> {
+    pub fn new(
+        buf: &'static mut heapless::Vec<u8, BLOCK_SIZE>,
+        write_block_fn: fn(
+            user_data: &T,
```

Comment:
It might be worthwhile to extract `WriteBlockFn` and `FinalizeFn` types, then use them here and in the `struct` definition.